doc/rfc001-sum-types.md

```diff
@@ -0,0 +1,846 @@
+sugar for a label typed with the unit (i.e., the empty record):
+
+```quint
+type S = C
+```
+
+Is normalized as
+
+```quint
+type S = C({})
+```
+
+
+#### Injection
+
+I propose introducing a lexical rule to syntactically differentiate data
+constructor of sum types from operator application: any identifiers starting
+with an uppercase letter that is occurs in the syntactic position of operator
+application is parsed as a sum type constructor: 
+
+```quint
+// Constructoin of a previously declared sum type
+val a : T = A(1)
+// Construction of a polymorphic variant of an anonymous type
+val b : AdHocVariant(int) | v = AdHocVariant(1)
+```
+
+Identifiers starting with a capital letter would them be available only for
+module names, type aliases, and variant constructors.
+
+This is the only way I can think of to preserve a Rust-like syntax for sum type
+injectors and also support the anonymous sum types that available via
+row-ploymorphic sum types. If we instead require all variant types to be defined
+in advance.
+
+In any case, I think there is a gain in readability and parseability in ensuring
+we reason about code purely locally to determine whether a normal operator is
+being applied or
+
+We could also have a normal form for injection called `inj` or `variant`.  If we
+follow our current convention of exotic operators that lift quint strings into
+labels, this would look like:
+
+```quint
+val a : T = inj("A", 1)
+```
+
+Or if we take my recommendation below, we'd have `inj(A, 1)`.
+
+
+#### Case analyses
+
+**Option 1**
+
+```quint
+match e {
+  | A(a) => ...
+  | B(b) => ...
+}
+```
+
+`|` can be swapped for `,`, but IMO it is valuable and clarifying to keep the
+syntactic connection with disjunction and with the type syntax. 
+
+The normal form can be `match(e, {...})` to match our current conventions. 
+
+##### As an anonymous operator
+
+We could also consider a case analysis `{ A : a => e1 | B : b => e2 }` -- where
+`e1, e2 : S` -- as syntax sugar for an anonymous operator of type `T => S`.
+`match` would then be syntax sugar for operator application. This follows
+Scala's case blocks or OCaml's `function` keyword and would allow idioms such
+as:
+
+```quint
+setOfTs.map({ A(a) => e1 | B(b) => e2 })
+```
+
+instead of requiring
+
+```quint
+setOfTs.map(x => match x { A(a) => e1 | B(b) => e2 })
+```
+
+### Statics
+
+These type rules assume we keep our current approach of using quint strings for
+labels. But see my argument for simplifying our approach under [Drop the exotic
+operators](#drop-the-exotic-operators). See the discussion in
+[Statics](#statics-1) below for a detailed explanation and analysis.
+
+#### Injection
+
+The typing rule for constructing a variant of a sum type:
+
+$$
+\frac
+{ \Gamma \vdash e \colon (t, c) \quad \Gamma \vdash \`l\` \colon str \quad fresh(s) }
+{ \Gamma \vdash \ inj(\`l\`, e) \ \colon (s, c \land s \sim \\{ \ l \colon t | tail\_s \ \\}) }
+$$
```

Comment:
VSCode preview produces a syntax error, most likely, on `\`l\``, but I could get the meaning.

This rule poses one interesting question. If we assume that the user specifies the type beforehand, e.g., via:

```
type T =
  | A(int)
  | B(str)
```

then the type of `T` should be closed, that is, it does not contain a row variable. There are exactly two variables. I do not see how it is reflected in the above rule. If the above rule assumes open types, then we would be able to mix constructors of different sum types. I don't think it is the expected behavior, e.g., in Scala.

Reply:
this was indeed following the rules for row-polymorphic variants, and constructing a variant in that system leaves the row open. e.g., in utop:

```ocaml
utop # let x = `A(1);;
val x : [> `A of int ] = `A 1
utop # let y : [`B | `A of int] = x;;
val y : [ `A of int | `B ] = `A 1
```

You can freely mix variants so long as they fit structurally with one of the alternatives. In the context where you are working with a closed type, as in the annotated `y` above, you can only use a constructor that is occurs in the alternatives.

Since we are not aiming to take advantage of the row-types any more, as per our discussion, I'll need to tweak the rule.